source/common/common/assert.cc

```diff
@@ -28,15 +28,66 @@ class ActionRegistrationImpl : public ActionRegistration {
   static std::function<void()> debug_assertion_failure_record_action_;
 };
 
+class EnvoyBugRegistrationImpl : public ActionRegistration {
+public:
+  EnvoyBugRegistrationImpl(std::function<void()> action) {
+    ASSERT(envoy_bug_failure_record_action_ == nullptr,
+           "An ENVOY_BUG action was already set. Currently only a single action is supported.");
+    count_ = 0;
+    envoy_bug_failure_record_action_ = action;
+  }
+
+  ~EnvoyBugRegistrationImpl() override {
+    ASSERT(envoy_bug_failure_record_action_ != nullptr);
+    envoy_bug_failure_record_action_ = nullptr;
+  }
+
+  static bool shouldLogAndInvoke() {
```

Comment:
Regarding exponential backoff...

I think we want a separate counter for each ENVOY_BUG instance.  I think that the macro needs to instantiate a static in the macro which owns the static count that is applicable to that ENVOY_BUG invocation and controls the exponential backoff.

Without that, more frequent bugs will make less frequent ones undetectable.